### PR TITLE
NETOBSERV-1805: threads are leaking with continous adding and deleting pods [BP 1.7]

### PR DESCRIPTION
**Pull request by msherif1234**
Signed-off-by: Mohamed Mahmoud <mmahmoud@redhat.com>
(cherry picked from commit 6111b9846d226a1478490514430807be86ca6e44)

## Dependencies

n/a

## Checklist

If you are not familiar with our processes or don't know what to answer in the list below, let us know in a comment: the maintainers will take care of that.

* [ ] Will this change affect NetObserv / Network Observability operator? If not, you can ignore the rest of this checklist.
* [ ] Is this PR backed with a JIRA ticket? If so, make sure it is written as a title prefix _(in general, PRs affecting the NetObserv/Network Observability product should be backed with a JIRA ticket - especially if they bring user facing changes)._
* [ ] Does this PR require product documentation?
  * [ ] If so, make sure the JIRA epic is labelled with "documentation" and provides a description relevant for doc writers, such as use cases or scenarios. Any required step to activate or configure the feature should be documented there, such as new CRD knobs.
* [ ] Does this PR require a product release notes entry?
  * [ ] If so, fill in "Release Note Text" in the JIRA.
* [ ] Is there anything else the QE team should know before testing? E.g: configuration changes, environment setup, etc.
  * [ ] If so, make sure it is described in the JIRA ticket.
* QE requirements (check 1 from the list):
  * [ ] Standard QE validation, with pre-merge tests unless stated otherwise.
  * [ ] Regression tests only (e.g. refactoring with no user-facing change).
  * [ ] No QE (e.g. trivial change with high reviewer's confidence, or per agreement with the QE team).

### CODE REVIEW

File: pkg/ifaces/watcher.go [hunk 1]
```diff
@@ -191,9 +195,15 @@ func (w *Watcher) netnsNotify(ctx context.Context, out chan Event) {
 				}
 				if event.Op&fsnotify.Create == fsnotify.Create {
 					ns := filepath.Base(event.Name)
-					log.WithField("netns", ns).Debug("netns notification")
+					log.WithField("netns", ns).Debug("netns create notification")
+					w.nsDone[ns] = make(chan struct{})
```
Comment (jotak) [inline pkg/ifaces/watcher.go:199]: when creating a new channel for map key `ns`, is there any risk of collision / leak if there was already previously a channel stored in `w.nsDone[ns]` ? Might be safer to first check if it isn't empty, close the old channel ?

Comment (msherif1234) [reply pkg/ifaces/watcher.go:199]: I think that will be an issue as we suppose to know about netns creation or deletion once

File: pkg/ifaces/watcher.go [hunk 2]
```diff
@@ -31,6 +31,7 @@ type Watcher struct {
 	linkSubscriberAt func(ns netns.NsHandle, ch chan<- netlink.LinkUpdate, done <-chan struct{}) error
 	mutex            *sync.Mutex
 	netnsWatcher     *fsnotify.Watcher
+	nsDone           map[string]chan struct{}
```
[the remark this answers is not in the document]
Comment (msherif1234) [reply pkg/ifaces/watcher.go:34]: I added defer in sendUpdates to close them

File: pkg/ifaces/watcher.go [hunk 3]
```diff
@@ -191,9 +199,19 @@ func (w *Watcher) netnsNotify(ctx context.Context, out chan Event) {
...
+						log.WithField("netns", ns).Debug("netns channel already exists, delete it")
+						delete(w.nsDone, ns)
+					}
+					w.nsDone[ns] = make(chan struct{})
 					go w.sendUpdates(ctx, ns, out)
 				}
+				if event.Op&fsnotify.Remove == fsnotify.Remove {
+					ns := filepath.Base(event.Name)
+					log.WithField("netns", ns).Debug("netns delete notification")
+					w.nsDone[ns] <- struct{}{}
```
Comment (jotak) [inline pkg/ifaces/watcher.go:213]: there is a risk that this code is called after the channel was closed and deleted?
ie. `w.sendUpdates` finishes, and then we get a fsnotify.Remove?
(just asking)

Comment (msherif1234) [reply pkg/ifaces/watcher.go:213]: the code in sendUpdates is blocking won't finish till done is happen which is triggered via fsnotify

Comment (msherif1234) [reply pkg/ifaces/watcher.go]: I added a check around the remove to be safe

**Pull request merged**